trollmoves/tests/test_client.py

```diff
@@ -701,98 +751,223 @@ def test_reload_config(Listener, NoisyPublisher):
+    from trollmoves.client import reload_config
+
+    chains = {}
+    callback = MagicMock()
+
+    try:
+        reload_config(client_config_1_item, chains, callback=callback)
+        num_providers = len(chains["eumetcast_hrit_0deg_scp_hot_spare"]._config['providers'])
+        assert Listener.call_count == num_providers
+        reload_config(client_config_1_item_non_pub_provider_item_modified, chains, callback=callback)
+        assert Listener.call_count == num_providers
+    finally:
+        for key in chains:
+            try:
+                chains[key].stop()
+            except AttributeError:
+                pass
+        os.remove(client_config_1_item)
+        os.remove(client_config_1_item_non_pub_provider_item_modified)
+
+
+@patch('trollmoves.client.NoisyPublisher')
+@patch('trollmoves.client.Listener')
+def test_reload_config_providers_added(Listener, NoisyPublisher, client_config_1_item,
+                                       client_config_1_item_two_providers):
+    """Test trollmoves.client.reload_config() when providers are added."""
+    from trollmoves.client import reload_config
+
+    chains = {}
+    callback = MagicMock()
+
+    try:
+        reload_config(client_config_1_item_two_providers, chains, callback=callback)
+        num_providers = len(chains["eumetcast_hrit_0deg_scp_hot_spare"]._config['providers'])
+        assert len(chains["eumetcast_hrit_0deg_scp_hot_spare"].listeners) == num_providers
+        assert Listener.call_count == num_providers
+        reload_config(client_config_1_item, chains, callback=callback)
+        num_providers2 = len(chains["eumetcast_hrit_0deg_scp_hot_spare"]._config['providers'])
+        assert len(chains["eumetcast_hrit_0deg_scp_hot_spare"].listeners) == num_providers2
+        assert Listener.call_count == num_providers2
+    finally:
+        for key in chains:
+            try:
+                chains[key].stop()
+            except AttributeError:
+                pass
+        os.remove(client_config_1_item_two_providers)
+        os.remove(client_config_1_item)
+
+
+@patch('trollmoves.client.NoisyPublisher')
+@patch('trollmoves.client.Listener')
+def test_reload_config_providers_removed(Listener, NoisyPublisher, client_config_1_item,
+                                         client_config_1_item_two_providers):
+    """Test trollmoves.client.reload_config() when providers are removed."""
+    from trollmoves.client import reload_config
+
+    listener = MagicMock()
+    Listener.return_value = listener
+
+    chains = {}
+    callback = MagicMock()
+
+    try:
+        reload_config(client_config_1_item, chains, callback=callback)
+        num_providers = len(chains["eumetcast_hrit_0deg_scp_hot_spare"]._config['providers'])
+        assert len(chains["eumetcast_hrit_0deg_scp_hot_spare"].listeners) == num_providers
+        assert Listener.call_count == num_providers
+        reload_config(client_config_1_item_two_providers, chains, callback=callback)
```

Comment:
There is some code duplication between this function and the next.

Reply:
Refactored the provider tests, and all the places where chains are stopped.